### PR TITLE
upgrade testing supports

### PR DESCRIPTION
**Pull request by turadg**
refs: #9303

## Description

A bunch of refactors and testing utilities to support #9303.

It also adds a couple upgrade tests of orchestration contracts, but neither of them cover restarting orchestration flows so I'll leave #9755 for that. 

### Security Considerations
none

### Scaling Considerations
none

### Documentation Considerations
none

### Testing Considerations
per se

### Upgrade Considerations
Helps to test upgrades, should not affect any.

### CODE REVIEW

File: packages/vats/src/core/types-ambient.d.ts [hunk 1]
```diff
@@ -371,8 +371,10 @@ type ChainBootstrapSpaceT = {
   namesByAddress: import('../types.js').NameHub;
   namesByAddressAdmin: import('../types.js').NamesByAddressAdmin;
   networkVat: NetworkVat;
+  orchestration?: CosmosInterchainService;
```
Comment (0xpatrickdev) [inline packages/vats/src/core/types-ambient.d.ts:374]:
```suggestion
  cosmosInterchainService?: CosmosInterchainService;
  orchestrationVat?: OrchestrationVat;
```
In light of #9668

File: packages/boot/tools/supports.ts [hunk 2]
```diff
@@ -22,21 +22,55 @@ import { initSwingStore } from '@agoric/swing-store';
...
+  name: N,
+) => N extends keyof EconomyBootstrapPowers['consume']
+  ? EconomyBootstrapPowers['consume'][N]
+  : unknown;
+
+// XXX should satisfy EVProxy from run-utils.js but that's failing to import
+/**
+ * Elaboration of EVProxy with knowledge of bootstrap space in these tests.
+ */
+type BootstrapEV = EProxy & {
```
Comment (dckc) [inline packages/boot/tools/supports.ts:61]: .ts :ninja:

File: packages/orchestration/src/examples/basic-flows.contract.js [hunk 3]
```diff
@@ -42,13 +42,14 @@ const makeOrchAccountHandler = async (orch, _ctx, seat, { chainName }) => {
...
- * @param {MakePortfolioHolder} makePortfolioHolder
+ * @param {object} ctx
+ * @param {MakePortfolioHolder} ctx.makePortfolioHolder
  * @param {ZCFSeat} seat
  * @param {{ chainNames: string[] }} offerArgs
  */
 const makePortfolioAcctHandler = async (
   orch,
-  makePortfolioHolder,
+  { makePortfolioHolder },
```
Comment (0xpatrickdev) [inline packages/orchestration/src/examples/basic-flows.contract.js]: Had this in mind with the original implementation: https://github.com/Agoric/agoric-sdk/pull/9591#discussion_r1657946588

But that is platform code and this is an example contract. So, don't see any harm here

File: packages/zoe/src/contractFacet/types-ambient.d.ts [hunk 4]
```diff
@@ -209,7 +209,7 @@ type ZCFSeat = import('@endo/pass-style').RemotableObject & {
 };
 type ZcfSeatKit = {
   zcfSeat: ZCFSeat;
-  userSeat: ERef<UserSeat>;
+  userSeat: Promise<UserSeat>;
```
Comment (dckc) [inline packages/zoe/src/contractFacet/types-ambient.d.ts:212]: This excludes a remote presence for a UserSeat. That doesn't seem right. But I guess it could be a promise, so any consumers have to await it anyway. so close enough.

File: packages/orchestration/src/examples/basic-flows.flows.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+ * @param {any} _ctx
+ * @param {ZCFSeat} seat
+ * @param {{ chainName: string }} offerArgs
+ */
+export const makeOrchAccount = async (orch, _ctx, seat, { chainName }) => {
+  seat.exit(); // no funds exchanged
+  mustMatch(chainName, M.string());
+  const remoteChain = await orch.getChain(chainName);
+  const cosmosAccount = await remoteChain.makeAccount();
+  return cosmosAccount.asContinuingOffer();
```
Comment (0xpatrickdev) [inline packages/orchestration/src/examples/basic-flows.flows.js]: If you find yourself needing to make fixups, mind including this drive-by?

```diff
- * Create an account on a Cosmos chain and return a continuing offer with
- * invitations makers for Delegate, WithdrawRewards, Transfer, etc.
+ * Create an OrchestrationAccount for a specific chain and return a continuing
+ * offer with invitations makers for Delegate, WithdrawRewards, Transfer, etc.
  *
  * @satisfies {OrchestrationFlow}
  * @param {Orchestrator} orch
@@ -30,8 +30,8 @@ const makeOrchAccountHandler = async (orch, _ctx, seat, { chainName }) => {
   seat.exit(); // no funds exchanged
   mustMatch(chainName, M.string());
   const remoteChain = await orch.getChain(chainName);
-  const cosmosAccount = await remoteChain.makeAccount();
-  return cosmosAccount.asContinuingOffer();
+  const orchAccount = await remoteChain.makeAccount();
+  return orchAccount.asContinuingOffer();
 };
 ```

File: packages/orchestration/src/examples/basic-flows.contract.js [hunk 6]
```diff
@@ -89,23 +22,18 @@ const makePortfolioAcctHandler = async (
...
-    undefined,
-    makeOrchAccountHandler,
-  );
-
-  const makePortfolioAccount = orchestrate(
-    'makePortfolioAccount',
-    { makePortfolioHolder },
-    makePortfolioAcctHandler,
-  );
+  const orchFns = orchestrateAll(flows, { makePortfolioHolder });
```
Comment (0xpatrickdev) [inline packages/orchestration/src/examples/basic-flows.contract.js:36]: These have different contexts - so it seems like 2x `orchestrate` is worth the cost here? Seems like it'll also be good to have examples using both `orchestrateAll` and `orchestrate`.

File: packages/boot/tools/supports.ts [hunk 7]
```diff
@@ -40,7 +41,7 @@ import type { MsgDelegateResponse } from '@agoric/cosmic-proto/cosmos/staking/v1
 import type { CoreEvalSDKType } from '@agoric/cosmic-proto/swingset/swingset.js';
 import type { EconomyBootstrapPowers } from '@agoric/inter-protocol/src/proposals/econ-behaviors.js';
 import type { SwingsetController } from '@agoric/swingset-vat/src/controller/controller.js';
-import type { BridgeHandler, IBCMethod } from '@agoric/vats';
+import type { BridgeHandler, IBCMethod, IBCPacket } from '@agoric/vats';
```
Comment (dckc) [inline packages/boot/tools/supports.ts:44]: The vats package is really bursting at the seams, isn't it?

I wonder what it would cost to move IBCMethod, IBCPacket to network.

File: packages/boot/tools/supports.ts [hunk 8]
```diff
@@ -437,7 +437,13 @@ export const makeSwingsetTestKit = async (
           case 'IBC_METHOD':
             switch (obj.method) {
               case 'startChannelOpenInit':
-                inbound!(BridgeId.DIBC, icaMocks.channelOpenAck(obj));
+                // This is what it used to do but that's not valid because bridge responses
+                // aren't instantaneous.
+                // inbound(BridgeId.DIBC, icaMocks.channelOpenAck(obj));
```
Comment (dckc) [inline packages/boot/tools/supports.ts]: we want to keep this?

File: packages/boot/tools/supports.ts [hunk 9]
```diff
@@ -327,12 +327,24 @@ export const makeSwingsetTestKit = async (
...
     ibcSequenceNonce += 1;
     const msg = icaMocks.ackPacket(obj, ibcSequenceNonce, ack);
-    inbound(BridgeId.DIBC, msg);
+    setTimeout(() => {
+      /**
+       * Mock when Agoric receives the ack from another chain over DIBC. Always
+       * happens after the packet is returned.
+       */
+      inbound(BridgeId.DIBC, msg);
+    });
```
Comment (0xpatrickdev) [inline packages/boot/tools/supports.ts:356]: 👍 nice use of setTimeout here

File: packages/boot/tools/supports.ts [hunk 10]
```diff
@@ -327,12 +327,24 @@ export const makeSwingsetTestKit = async (
...
+   * Finally return the packet that will be sent.
+   */
   const makeAckEvent = (obj: IBCMethod<'sendPacket'>, ack: string) => {
     ibcSequenceNonce += 1;
     const msg = icaMocks.ackPacket(obj, ibcSequenceNonce, ack);
-    inbound(BridgeId.DIBC, msg);
+    setTimeout(() => {
+      /**
+       * Mock when Agoric receives the ack from another chain over DIBC. Always
+       * happens after the packet is returned.
```
Comment (0xpatrickdev) [inline packages/boot/tools/supports.ts:353]: Should we also make a similar change in:
- packages/orchestration/test/network-fakes.ts
- packages/vats/tools/fake-bridge.js

File: packages/boot/test/orchestration/restart-contracts.test.ts [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+    },
+    proposal: {},
+  });
+  // no result yet because the IBC incoming messages haven't arrived
+  // and won't until we flush.
+  t.false(hasResult(wd.getLatestUpdateRecord()));
+
+  t.log('restart stakeAtom');
+  await evalProposal(
+    buildProposal('@agoric/builders/scripts/testing/restart-stakeAtom.js'),
```
Comment (dckc) [inline packages/boot/test/orchestration/restart-contracts.test.ts:164]: :clap:

File: packages/boot/test/bootstrapTests/orchestration.test.ts [hunk 12]
```diff
@@ -190,6 +197,7 @@ test.serial('stakeAtom - smart wallet', async t => {
     encoding: 'bech32',
   };
 
+  // This will trigger the immediate ack of the mock bridge
```
Comment (0xpatrickdev) [inline packages/boot/test/bootstrapTests/orchestration.test.ts:201]: ~~Is "This" related to `executeOffer`?~~ It seems to be due to `ackImmediately`. Please consider making this comment more explicit

File: packages/builders/scripts/testing/restart-sendAnywhere.js [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+export const defaultProposalBuilder = async () =>
+  harden({
+    // Somewhat unorthodox, source the exports from this builder module
+    sourceSpec: '@agoric/builders/scripts/testing/restart-sendAnywhere.js',
+    getManifestCall: [getManifest.name],
+  });
+
+export default async (homeP, endowments) => {
+  // import dynamically so the module can work in CoreEval environment
+  const dspModule = await import('@agoric/deploy-script-support');
```
Comment (dckc) [inline packages/builders/scripts/testing/restart-sendAnywhere.js:96]: interesting... I wonder how this doesn't run afoul of the SES prohibition on dynamic import.

**Review by dckc — APPROVED** [review]
very nice.

I'm not 100% confident that I grok "feat: defer inbound bridge messages" fully but I suppose the tests show that it works.

File: packages/zoe/src/contractFacet/types-ambient.d.ts (hunk 4 above)
Comment (turadg) [reply packages/zoe/src/contractFacet/types-ambient.d.ts:212]: `ERef` is just `T | Promise<T>` and it's never just `T`. The remoteness is handled elsewhere.

File: packages/boot/tools/supports.ts (hunk 8 above)
Comment (turadg) [reply packages/boot/tools/supports.ts]: just for code review. I'll drop it

File: packages/boot/test/bootstrapTests/orchestration.test.ts [hunk 14]
```diff
@@ -170,7 +175,9 @@ test.serial('stakeAtom - smart wallet', async t => {
   const { ATOM } = agoricNamesRemotes.brand;
   ATOM || Fail`ATOM missing from agoricNames`;
 
-  await wd.executeOffer({
+  // Cannot await executeOffer because the offer won't resolve until after the bridge's inbound queue is flushed.
+  // But this test doesn't require that.
```
Comment (0xpatrickdev) [inline packages/boot/test/bootstrapTests/orchestration.test.ts:180]: Should we require this test to check the offer result?

File: packages/builders/scripts/testing/restart-sendAnywhere.js (hunk 13 above)
Comment (turadg) [reply packages/builders/scripts/testing/restart-sendAnywhere.js:96]: Maybe `agoric run` isn't locked down so much?

File: packages/builders/scripts/testing/restart-stakeAtom.js [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+import { deeplyFulfilledObject, makeTracer } from '@agoric/internal';
+import { makeStorageNodeChild } from '@agoric/internal/src/lib-chainStorage.js';
+import { E } from '@endo/far';
+
+/// <reference types="@agoric/vats/src/core/types-ambient"/>
+
+const trace = makeTracer('RestartSA', true);
+
+/**
+ * @import {start as StartFn} from '@agoric/orchestration/src/examples/stakeIca.contract.js';
```
Comment (0xpatrickdev) [inline packages/builders/scripts/testing/restart-stakeAtom.js:14]: Given the refs, noting that `stakeIca` uses Vows but not `async-flow`

**Review by 0xpatrickdev — APPROVED** [review]

File: packages/vats/src/core/types-ambient.d.ts (hunk 1 above)
Comment (turadg) [reply packages/vats/src/core/types-ambient.d.ts:374]: That's a bigger change than just the typedef

File: packages/orchestration/src/examples/basic-flows.contract.js (hunk 6 above)
Comment (turadg) [reply packages/orchestration/src/examples/basic-flows.contract.js:36]: I vacillated on this but I don't see how providing `makePortfolioHolder` to both is a risk.

I think `orchestrateAll` is what we should steer people to in general because it gets the durable from a named export, which changing is known to break downstream, whereas the string argument was easily confused as a debugging name.

File: packages/builders/scripts/testing/restart-sendAnywhere.js (hunk 13 above)
Comment (dckc) [reply packages/builders/scripts/testing/restart-sendAnywhere.js:96]: But this whole module is bundled for use as the actual core-eval.

I suppose the `import()` gets "de-fanged" by deploy-script-support. It would probably fail if we tried to run this part of the code on-chain, but we don't.

File: packages/boot/tools/supports.ts (hunk 9 above)
Comment (dckc) [reply packages/boot/tools/supports.ts:356]: ambient authority grumble...

File: packages/boot/tools/supports.ts (hunk 10 above)
Comment (turadg) [reply packages/boot/tools/supports.ts:353]: I suppose so. WDYT @michaelfig ?

File: packages/boot/test/bootstrapTests/orchestration.test.ts (hunk 12 above)
Comment (turadg) [reply packages/boot/test/bootstrapTests/orchestration.test.ts:201]: I tried to word it in a way that wouldn't be false if that code was refactored

File: packages/boot/test/bootstrapTests/orchestration.test.ts (hunk 14 above)
Comment (turadg) [reply packages/boot/test/bootstrapTests/orchestration.test.ts:180]: that's covered in `restart-contracts`.

File: packages/builders/scripts/testing/restart-stakeAtom.js (hunk 15 above)
Comment (turadg) [reply packages/builders/scripts/testing/restart-stakeAtom.js:14]: right. that's what I was getting at in the PR description,
> but neither of them cover restarting orchestration flows so I'll leave https://github.com/Agoric/agoric-sdk/pull/9755 for that.

File: packages/vats/src/core/types-ambient.d.ts (hunk 1 above)
Comment (0xpatrickdev) [reply packages/vats/src/core/types-ambient.d.ts:374]: I thought we don't have `orchestration` in chain bootstrap space anymore? But, these two are there?

File: packages/orchestration/src/examples/basic-flows.flows.js (hunk 5 above)
Comment (turadg) [reply packages/orchestration/src/examples/basic-flows.flows.js:31]: I'll do that in the next PR

File: packages/orchestration/src/examples/basic-flows.contract.js (hunk 6 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/examples/basic-flows.contract.js:36]: 👍 the second argument is convincing

File: packages/boot/test/bootstrapTests/orchestration.test.ts (hunk 14 above)
Comment (0xpatrickdev) [reply packages/boot/test/bootstrapTests/orchestration.test.ts:180]: Thanks, seeing that now. Maybe we can combine these files at some point in the future so it's easier to follow

File: packages/boot/tools/supports.ts (hunk 9 above)
Comment (turadg) [reply packages/boot/tools/supports.ts:356]: ambient authority in a _mock_.
